### PR TITLE
Automate version increment for /azure-iotedge 

### PR DESCRIPTION
**Pull request by yophilav**
Introduce an automation for /azure-iotedge version file updates for release automation.
The script verifies and updates the following files for the last step of the release process:
For release/1.1:
- Azure/azure-iotedge:/latest-iotedge-lts.json

For release/1.2:
- latest-aziot-edge.json
- latest-aziot-identity-service.json

## Azure IoT Edge PR checklist:

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines and Best Practices
- [x] I have read the [contribution guidelines](https://github.com/azure/iotedge#contributing).
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request includes a concise summary of the enhancement or bug fix.

### Testing Guidelines
- [x] Pull request includes test coverage for the included changes.
- Description of the pull request includes 
	- [x] concise summary of tests added/modified
	- [x] local testing done.  

### Draft PRs
- Open the PR in `Draft` mode if it is:
	- Work in progress or not intended to be merged.
	- Encountering multiple pipeline failures and working on fixes.

_Note: We use the kodiakhq bot to merge PRs once the necessary checks and approvals are in place. When it merges a PR, kodiakhq converts the PR title to the commit title, PR description to the commit description, and squashes all the commits in the PR to a single commit. The net effect is that entire PR becomes a single commit. Please follow the best practices mentioned [here](https://chris.beams.io/posts/git-commit/#:~:text=The%20seven%20rules%20of%20a%20great%20Git%20commit,what%20and%20why%20vs.%20how%20For%20example%3A%20) for the PR title and description_

### CODE REVIEW

**Review by and-rewsmith — COMMENTED** [review]
Some things to consider:
1. I think it would help review process to add more PR description for what this is, how it fits into other automation, how we expect to use it.
2. I feel like one of my learnings from making the nested agent lock script in bash is that it is hard to maintain, and that next time I would highly prefer doing complicated logic in a C# application. If you feel like it is getting complicated it may be worth switching to C#.
3. Squashing dev commits before reviewers leave a bunch of comments will partially mitigate github ui from collapsing some discussions.

File: scripts/linux/github/updateLatestVersion.sh [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+    check_required_variables
+
+    [[ -z "$AZURE_IOTEDGE_REPO_PATH" ]] && { echo "\$IOTEDGE_REPO_PATH is undefined"; exit 1; }
+
+    branchVersion=$(echo ${BRANCH_NAME##*/})
+    latestReleasedVersion=$(get_latest_release_per_branch_name)
+
+    if [ "$BRANCH_NAME" == "refs/heads/release/1.1" ]; then
+        # Set target version file to be updated
+        TARGET_IE_FILE="$AZURE_IOTEDGE_REPO_PATH/latest-iotedge-lts.json"
```
[the remark this answers is not in the document]
Comment (and-rewsmith) [reply scripts/linux/github/updateLatestVersion.sh:155]: Ah found it in the root of this repo:
https://github.com/Azure/azure-iotedge

File: scripts/linux/github/updateLatestVersion.sh [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+        echo "I'm pretty sure you don't want to release from the main branch."
+        exit 1;
+    else
+        echo "Oh dear, how did you get here?!?"
+        echo "Let me not let you do the release from your pull request branch"
+        exit 1;
+    fi
+}
+
+# TODO: Integrate the script with the execution of the pipeline for the release page.
```
Comment (and-rewsmith) [inline scripts/linux/github/updateLatestVersion.sh:227]: Which pipeline are you suggesting to integrate it into?

**Comment by and-rewsmith** [comment]
Any chance we could get a terminal output to see how running the script looks?

File: scripts/linux/github/updateLatestVersion.sh (hunk 2 above)
Comment (yophilav) [reply scripts/linux/github/updateLatestVersion.sh:227]: Edgelet release pipeline.

**Comment by yophilav** [comment]
> Any chance we could get a terminal output to see how running the script looks?

Gotta wait for the demo ;)

File: scripts/linux/github/updateLatestVersion.sh [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        contentIe=$(cat $TARGET_IE_FILE)
+        latestEdgeletVersion=$(echo $contentIe | jq '."aziot-edge"' | tr -d '"')
+        contentIis=$(cat $TARGET_IIS_FILE)
+        latestIisVersion=$(echo $contentIis | jq '."aziot-identity-service"' | tr -d '"')
+
+        # Verify
+        echo "Sanity check iotedge version"
+        version_sanity_check $proposedEdgeletVersion $latestEdgeletVersion
+        echo "Sanity check docker image version"
+        version_sanity_check $proposedImageVersion get_latest_release_per_branch_name
```
Comment (damonbarry) [inline scripts/linux/github/updateLatestVersion.sh:200]: We're doing the sanity check for the Docker image version here. But We don't write the version anywhere (`latest-aziot-edge.json` only tracks the edgelet version). Do you know what that is?

Comment (yophilav) [reply scripts/linux/github/updateLatestVersion.sh:200]: I'm not sure. @Gordonwang0 any thoughts?

Comment (gordonwang0) [reply scripts/linux/github/updateLatestVersion.sh:200]: No, I don't know why we do this.

Comment (yophilav) [reply scripts/linux/github/updateLatestVersion.sh:200]: My wildest recollection was that it has to do with Diagnostic module and how it basically uses aziot version to determine the edgeAgent & edgeHub version. But I'm not super familiar with that portion of code to take a look.

**Comment by yophilav** [comment]
Closing this PR and turn them into Draft PR. As I further along in the integration process, there are some more adjustments I want to add to this script.

File: scripts/linux/github/updateLatestVersion.sh (hunk 3 above)
Comment (yophilav) [reply scripts/linux/github/updateLatestVersion.sh:200]: @damonbarry After talking with Varun, it seems that we currently don't have code that checks if the latest EA/EH version are running.

File: scripts/linux/github/updateLatestVersion.sh [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    [[ -z "$AZURE_IOTEDGE_REPO_PATH" ]] && { echo "\$IOTEDGE_REPO_PATH is undefined"; exit 1; }
+    [[ -z "$VERSION" ]] && { echo "\$VERSION is undefined"; exit 1; }
+    [[ -z "$GITHUB_PAT" ]] && { echo "\$GITHUB_PAT is undefined"; exit 1; }
+
+    cd $AZURE_IOTEDGE_REPO_PATH
+    git config user.name iotedge1
+    git commit -am "Prepare for Release $VERSION"
+    lastCommitHash=$(git log -n 1 --pretty=format:"%H")
+    git tag "$VERSION" $lastCommitHash
+    git push https://$GITHUB_PAT@github.com/Azure/azure-iotedge.git
```
Comment (damonbarry) [inline scripts/linux/github/updateLatestVersion.sh:255]: We should find a more secure way to do this (where the PAT isn't part of the URL). Soon.

**Review by damonbarry — APPROVED** [review]

**Pull request merged**